src/csm_testing/tests/sat_functional/hwinv/test_hwinv.py

```diff
@@ -0,0 +1,389 @@
+    ],
+    "node bmcs": [
+        'xname', 'FRUID', 'Manufacturer', 'Model', 'Part Number', 'SKU', 'Serial Number',
+        'Manager Type', 'Firmware Version'
+    ],
+    "router bmcs": [
+        'xname', 'FRUID', 'Manufacturer', 'Model', 'Part Number', 'SKU', 'Serial Number',
+        'Manager Type', 'Firmware Version'
+    ],
+    "mgmt switches": [
+        'xname', 'FRUID', 'Manufacturer', 'Model', 'Part Number', 'SKU', 'Serial Number',
+        'Chassis Type'
+    ],
+    "cabinet pdus": [
+        'xname', 'FRUID', 'Manufacturer', 'Model', 'Part Number', 'SKU', 'Serial Number',
+        'Equipment Type', 'Firmware Version'
+    ],
+    "cabinet pdu power connectors": [
+        'xname', 'FRUID', 'Manufacturer', 'Model', 'Part Number', 'SKU', 'Serial Number',
+        'Nominal Voltage', 'Outlet Type', 'Phase Wiring Type', 'Power Enabled',
+        'Rated Current Amps', 'Voltage Type'
+    ]
+}
+
+
+def get_present_components() -> List[str]:
+    """Retrieve the list of components present in the system using HSM."""
+    command = "cray hsm inventory hardware list --format json"
+    proc = subprocess.run(shlex.split(command), stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+    hsm_output = proc.stdout.decode()
+    components = json.loads(hsm_output)
+    return list(set([component["Type"] for component in components]))
+
+
+def adjust_expected_header(header: List[str], sat_output: str) -> List[str]:
+    """Return a copy of the specified header, removing keys which were omitted in the output"""
+    new_header = header.copy()
+    for col in header:
+        info_miss = f"INFO: All values for '{col}' are 'MISSING', omitting key."
+        info_empty = f"INFO: All values for '{col}' are 'EMPTY', omitting key."
+        if any(msg in sat_output for msg in (info_miss, info_empty)):
+            new_header.remove(col)
+    return new_header
+
+
+def get_column_names_list(input_string: str) -> List[str]:
+    """Return the list of column names specified in the header line of the input string."""
+    lines = input_string.strip().split('\n')
+    for line in lines:
+        if line.strip().startswith('| xname'):
+            header_line = line
+            return [col.strip() for col in header_line.split('|') if col.strip()]
+    return []
+
+
+class TestHwinv(unittest.TestCase):
+
+    def run_hwinv_test(self, command: str, expected_key: str) -> None:
+        """Helper function to run an hwinv command test."""
+        proc = subprocess.run(shlex.split(command), stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+        stdout = proc.stdout.decode()
+        stderr = proc.stderr.decode()
+        self.assertIn(f"Listing of all {expected_key}", stdout)
+        actual_header = get_column_names_list(stdout)
+        expected_header = adjust_expected_header(HWINV_HEADERS[expected_key], stderr)
+        if not actual_header:
+            return
```

Comment:
If there are components of the type, we should expect there to be a table with content. We could store the list of component types from HSM (e.g. in the `setUpClass` method), and then we can use the mapping that from `sat hwinv` section headers to HSM component types to figure out if there should be a table with content.

Note that the logic you have here is essentially also the same as the logic you want in the two `--list-all` test cases, but in those test cases, you are performing that logic multiple times, i.e., once for each section of the output.